Security_Engine/Compute/CameraFieldOfView.cs

```diff
@@ -0,0 +1,351 @@
+
+        [Description("Split line in given obstacles.")]
+        [Input("rayLine", "Line to split.")]
+        [Input("obstacles", "Polyline objects to split the line in.")]
+        [Output("splitLines", "Lines splited by the obstacles.")]
+        private static List<Line> SplitLinesInObstacles(this Line line, List<Polyline> obstacles)
+        {
+            List<Point> splitPoints = new List<Point>();
+            foreach (Polyline obstacle in obstacles)
+            {
+                foreach (Line obstLine in obstacle.SubParts())
+                {
+                    Point point = line.LineIntersection(obstLine);
+                    if (point != null)
+                        splitPoints.Add(point);
+                }
+            }
+            return line.SplitAtPoints(splitPoints);
+        }
+
+        /***************************************************/
+
+        [Description("Check if line is inside obstacle.")]
+        [Input("line", "Line to check if it's inside obstacle.")]
+        [Input("obstacles", "Polyline objects to check if line is inside.")]
+        [Output("bool", "Boolean result of the checking.")]
+        private static bool IsInsideObstacle(this Line line, List<Polyline> obstacles)
+        {
+            foreach (Polyline obstacle in obstacles)
+            {
+                foreach (Line obstLine in obstacle.SubParts())
+                {
+                    if (obstLine.IsOnCurve(line.Start) && obstLine.IsOnCurve(line.End))
+                        return false;
+                }
+                if (obstacle.IsContaining(new List<Point> { line.Centroid() }, true))
+                {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        /***************************************************/
+
+        [Description("Create visible line from rayline passing through obstacles.")]
+        [Input("obstacles", "Polyline objects that are needed to create visible line.")]
+        [Output("visibleLine", "Line that is visible for the camera view.")]
+        private static Line VisibleLine(this Line rayLine, List<Polyline> obstacles)
+        {
+            List<Line> splitLines = rayLine.SplitLinesInObstacles(obstacles);
+            List<Line> outsideLines = new List<Line>();
+            foreach (Line splitLine in splitLines)
+            {
+                if (!splitLine.IsInsideObstacle(obstacles))
+                    outsideLines.Add(splitLine);
+            }
+            //create visible line
+            Line visibleLine = new Line();
+            if (outsideLines.Count == 1)
+                visibleLine = outsideLines[0];
+            else
+            {
+                Point startPoint = outsideLines[0].Start;
+                Point endPoint = outsideLines[0].End;
+                for (int i = 1; i < outsideLines.Count; i++)
+                {
+                    if (endPoint.IsEqual(outsideLines[i].Start))
```

Comment:
You should be passing the tolerance value through the entire method chain, and for example here it should be passed too

Reply:
Agreed, tolerance has been added where possible.